### ISSUE TITLE
`rop=type:directx` in combination with `encoding=utf-8` slows down global command

### ISSUE THREAD

**Issue by markonm** [vim/vim#2715]
Os: Windows 10 x64 (v1709)
Vim version:
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 11 2018 23:08:09)
MS-Windows 32-bit GUI version with OLE support
Included patches: 1-1599
Compiled by appveyor@APPVYR-WIN
```

Steps to reproduce:

`gvim --clean +"set rop=type:directx encoding=utf-8" +"norm Itest" +"norm yy10000p" +"g/./norm Itest"`

7.5 s

`gvim --clean +"set encoding=utf-8" +"norm Itest" +"norm yy10000p" +"g/./norm Itest"`

0.7 s

:global command is ten times slower when directx is enabled.

**Comment by Sharlock93** [comment]
it isn't just a single command, everything about vim comes to a crawl when you set rendering to directx, if you are trying to test it, you must set ``encoding=utf-8`` for it to display anything.
The slow down makes vim practically unusable, open the help pages and set the rendering to directx and boom you can barely move around, truly a shame because it really helps with the quality of the rendering.

**Comment by k-takata** [comment]
When I changed the last command `+"g/./norm Itest"` to `+"g/./call setline('.', 'test')"` or `+g/./norm D`, the slowdown didn't happen even with directx.
I also measured the speed of the `edit()` function in `edit.c`. It took about 1 μs with directx and about 0.1 μs without directx. So it seems that editing (e.g. `norm I`) is ten times slower with directx.
I found that `gui_mch_flush()` is unnecessarily called several times from `edit()` and this is the cause of the slowdown.

The following patch works around the issue:

```diff
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -6687,6 +6687,9 @@ gui_mch_flush(void)
     BOOL  __stdcall GdiFlush(void);
 #   endif
 
+    if (global_busy)
+	return;
+
 #if defined(FEAT_DIRECTX)
     if (IS_ENABLE_DIRECTX())
 	DWriteContext_Flush(s_dwc);
```

**Comment by markonm** [comment]
Thank you @k-takata! I confirm the patch fixed the issue. It even makes DirectX faster than GDI when tested with steps from initial post.

**Comment by brammool** [comment]
gui_mch_flush() is not called directly from edit().  Is there a more generic solution than just making this exception for Win32 GUI?

**Comment by k-takata** [comment]
It seems that there are several paths to call gui_mch_flush() from edit(), but I haven't found the path which causes the issue yet...

**Comment by k-takata** [comment]
Okay, this is a new patch:

```diff
--- a/src/edit.c
+++ b/src/edit.c
@@ -8855,7 +8855,7 @@ ins_esc(
      */
     if (reg_recording != 0 || restart_edit != NUL)
 	showmode();
-    else if (p_smd)
+    else if (p_smd && !global_busy)
 	msg("");
 
     return TRUE;	    /* exit Insert mode */
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -3084,9 +3084,10 @@ inchar(
 
 	/*
 	 * Always flush the output characters when getting input characters
-	 * from the user.
+	 * from the user (except in the :global command).
 	 */
-	out_flush();
+	if (!global_busy)
+	    out_flush();
 
 	/*
 	 * Fill up to a third of the buffer, because each character may be
```

`gui_mch_flush()` was also called from `ex_normal()`, not only from `edit()`. The call sequences for `norm I` are:

1. ex_normal() → exec_normal_cmd() → exec_normal() → normal_cmd() → nv_edit() → invoke_edit() → edit() → ins_esc() → msg() → msg_attr_keep() → msg_end() → out_flush() → ... → gui_mch_flush()
2. ex_normal() → setmouse() → update_mouseshape() → char_avail() → vpeekc() → vgetorpeek() → inchar() → out_flush → ... → gui_mch_flush()

**Comment by k-takata** [comment]
Another way is:

```diff
--- a/src/edit.c
+++ b/src/edit.c
@@ -8855,7 +8855,7 @@ ins_esc(
      */
     if (reg_recording != 0 || restart_edit != NUL)
 	showmode();
-    else if (p_smd)
+    else if (p_smd && !global_busy)
 	msg("");
 
     return TRUE;	    /* exit Insert mode */
--- a/src/ex_docmd.c
+++ b/src/ex_docmd.c
@@ -10441,7 +10441,8 @@ ex_normal(exarg_T *eap)
     restore_current_state(&save_state);
     --ex_normal_busy;
 #ifdef FEAT_MOUSE
-    setmouse();
+    if (!global_busy)
+	setmouse();
 #endif
 #ifdef CURSOR_SHAPE
     ui_cursor_shape();		/* may show different cursor shape */
```

I'm not sure which is better.

**Comment by brammool** [comment]
Ken Takata wrote:

> Another way is:
> 
> ```diff
> --- a/src/edit.c
> +++ b/src/edit.c
> @@ -8855,7 +8855,7 @@ ins_esc(
>       */
>      if (reg_recording != 0 || restart_edit != NUL)
>  	showmode();
> -    else if (p_smd)
> +    else if (p_smd && !global_busy)
>  	msg("");

I don't think it is useful to do any mode messages while global_busy is
set.  There already is a condition at the start of showmode(), we could
extend it with a check for global_busy.

We should put that condition in a separate function and set a flag when
showmode() is skipped because global_busy is set.  Then when global_busy
is reset, check for that flag and call showmode() if needed.

The msg() call you quoted should also call that function.


>      return TRUE;	    /* exit Insert mode */
> --- a/src/ex_docmd.c
> +++ b/src/ex_docmd.c
> @@ -10441,7 +10441,8 @@ ex_normal(exarg_T *eap)
>      restore_current_state(&save_state);
>      --ex_normal_busy;
>  #ifdef FEAT_MOUSE
> -    setmouse();
> +    if (!global_busy)
> +	setmouse();
>  #endif
>  #ifdef CURSOR_SHAPE
>      ui_cursor_shape();		/* may show different cursor shape */
> ```
> 
> I'm not sure which is better.

How about a more generic solution: Check for global_busy() in
char_avail(): 

--- vim81/src/getchar.c       2019-01-20 15:30:36.905328616 +0100
+++ getchar.c   2019-01-24 13:04:47.555974806 +0100
@@ -1923,6 +1923,10 @@
     if (disable_char_avail_for_testing)
        return FALSE;
 #endif
+    if (global_busy)
+       // busy processing text, avoid being slowed down by checking for typed
+       // characters
+       return FALSE;
     ++no_mapping;
     retval = vpeekc();
     --no_mapping;

And skip flushing when only peeking:

@@ -3084,9 +3088,10 @@
 
        /*
         * Always flush the output characters when getting input characters
-        * from the user.
+        * from the user and not just peeking.
         */
-       out_flush();
+       if (wait_time == -1L || wait_time > 10L)
+           out_flush();
 


-- 
E  M  A  C  S
s  e  l  o  h
c  t  t  n  i
a  a     t  f
p        r  t
e        o
         l

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by k-takata** [comment]
> I don't think it is useful to do any mode messages while global_busy is set. There already is a condition at the start of showmode(), we could extend it with a check for global_busy.

> We should put that condition in a separate function and set a flag when showmode() is skipped because global_busy is set. Then when global_busy is reset, check for that flag and call showmode() if needed.

> The msg() call you quoted should also call that function.

Hmm, sorry, I couldn't understand well.

> How about a more generic solution: Check for global_busy() in char_avail():

This doesn't work well. It causes slowdown.

> And skip flushing when only peeking:

This works well.
The patch is now:

```diff
diff --git a/src/getchar.c b/src/getchar.c
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -3039,9 +3039,10 @@ inchar(
 
 	/*
 	 * Always flush the output characters when getting input characters
-	 * from the user.
+	 * from the user and not just peeking.
 	 */
-	out_flush();
+	if (wait_time == -1L || wait_time > 10L)
+	    out_flush();
 
 	/*
 	 * Fill up to a third of the buffer, because each character may be
diff --git a/src/message.c b/src/message.c
--- a/src/message.c
+++ b/src/message.c
@@ -135,6 +135,13 @@ msg_attr_keep(
     int		retval;
     char_u	*buf = NULL;
 
+    if (global_busy)
+    {
+	// busy processing text, redraw later.
+	redraw_cmdline = TRUE;
+	return TRUE;
+    }
+
     /* Skip messages not matching ":filter pattern".
      * Don't filter when there is an error. */
     if (!emsg_on_display && message_filtered((char_u *)s))
diff --git a/src/screen.c b/src/screen.c
--- a/src/screen.c
+++ b/src/screen.c
@@ -10149,7 +10149,8 @@ showmode(void)
 	 * Call char_avail() only when we are going to show something, because
 	 * it takes a bit of time.
 	 */
-	if (!redrawing() || (char_avail() && !KeyTyped) || msg_silent != 0)
+	if (!redrawing() || (char_avail() && !KeyTyped) || msg_silent != 0
+		|| global_busy)
 	{
 	    redraw_cmdline = TRUE;		/* show mode later */
 	    return 0;
```

**Comment by brammool** [comment]
Ken Takata wrote:

> > I don't think it is useful to do any mode messages while global_busy
> > is set. There already is a condition at the start of showmode(), we
> > could extend it with a check for global_busy.
> 
> > We should put that condition in a separate function and set a flag
> > when showmode() is skipped because global_busy is set. Then when
> > global_busy is reset, check for that flag and call showmode() if
> > needed.
> 
> > The msg() call you quoted should also call that function.
> 
> Hmm, sorry, I couldn't understand well.

There is a call to msg() with an empty string in an earlier version of
your patch.  It's to remove the mode message when returning from Insert
mode.  This only matters if the global command starts and ends Insert
mode, which would be unusual.  But if showmode() skips the output, this
one should too.

> > How about a more generic solution: Check for global_busy() in char_avail():
> 
> This doesn't work well. It causes slowdown.

I also suspect it might cause CTRL-C to not interrupt the global
command.  So let's not include this change.
 
> > And skip flushing when only peeking:
> 
> This works well.
> The patch is now:
> 
> ```diff
> diff --git a/src/getchar.c b/src/getchar.c
> --- a/src/getchar.c
> +++ b/src/getchar.c
> @@ -3039,9 +3039,10 @@ inchar(
>  
>  	/*
>  	 * Always flush the output characters when getting input characters
> -	 * from the user.
> +	 * from the user and not just peeking.
>  	 */
> -	out_flush();
> +	if (wait_time == -1L || wait_time > 10L)
> +	    out_flush();

OK, this should work.

> diff --git a/src/message.c b/src/message.c
> --- a/src/message.c
> +++ b/src/message.c
> @@ -135,6 +135,13 @@ msg_attr_keep(
>      int		retval;
>      char_u	*buf = NULL;
>  
> +    if (global_busy)
> +    {
> +	// busy processing text, redraw later.
> +	redraw_cmdline = TRUE;
> +	return TRUE;
> +    }
> +
>      /* Skip messages not matching ":filter pattern".

Hmm, if the global command intentionally produces messages, they would
be dropped.  I don't think we should do this.  As mentioned above, if
it's about surpressing the mode message (and the ruler), we should find
another way to skip that.

>       * Don't filter when there is an error. */
>      if (!emsg_on_display && message_filtered((char_u *)s))
> diff --git a/src/screen.c b/src/screen.c
> --- a/src/screen.c
> +++ b/src/screen.c
> @@ -10149,7 +10149,8 @@ showmode(void)
>  	 * Call char_avail() only when we are going to show something, because
>  	 * it takes a bit of time.
>  	 */
> -	if (!redrawing() || (char_avail() && !KeyTyped) || msg_silent != 0)
> +	if (!redrawing() || (char_avail() && !KeyTyped) || msg_silent != 0
> +		|| global_busy)
>  	{
>  	    redraw_cmdline = TRUE;		/* show mode later */
>  	    return 0;

Right, that is in showmode() itself.

I'll make a patch from this, we can make further improvements after
that.

-- 
A vacation is a period of travel during which you find that you
took twice as many clothes and half as much money as you needed.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by k-takata** [comment]
@brammool Thank you for making the patch.

Fixed in [8.1.0822](https://github.com/vim/vim/commit/cb574f415486adff645ce384979bfecf27f5be8c). Closing.
